src/zkp/pimod.rs

```diff
@@ -90,36 +122,44 @@ impl Proof for PiModProof {
         rng: &mut R,
     ) -> Result<Self> {
         // Step 1: Pick a random w in [1, N) that has a Jacobi symbol of -1
-        let mut w = random_positive_bn(rng, &input.N);
-        while jacobi(&w, &input.N) != -1 {
-            w = random_positive_bn(rng, &input.N);
+        let mut w = random_positive_bn(rng, &input.modulus);
+        while jacobi(&w, &input.modulus) != -1 {
+            w = random_positive_bn(rng, &input.modulus);
         }
 
         Self::fill_transcript(transcript, context, input, &w)?;
 
-        let mut elements = vec![];
-        for _ in 0..LAMBDA {
-            let y = positive_challenge_from_transcript(transcript, &input.N)?;
-            let (a, b, x) = y_prime_combinations(&w, &y, &secret.p, &secret.q)?;
-
-            // Compute phi(N) = (p-1) * (q-1)
-            let phi_n = (&secret.p - 1) * (&secret.q - 1);
-            let exp = input.N.invert(&phi_n).ok_or_else(|| {
-                error!("Could not invert N");
-                InternalError::InternalInvariantFailed
-            })?;
-            let z = modpow(&y, &exp, &input.N);
-
-            elements.push(PiModProofElements {
-                x: x[0].clone(),
-                a,
-                b,
-                z,
-                y,
-            });
-        }
-
-        let proof = Self { w, elements };
+        let elements = (0..LAMBDA)
+            .map(|_| {
+                let y = positive_challenge_from_transcript(transcript, &input.modulus)?;
+                let (a, b, mut x) = y_prime_combinations(&w, &y, &secret.p, &secret.q)?;
+
+                let phi_n = (&secret.p - 1) * (&secret.q - 1);
+                let exp = input.modulus.invert(&phi_n).ok_or_else(|| {
+                    error!("Could not invert N");
+                    InternalError::InternalInvariantFailed
+                })?;
+                let z = modpow(&y, &exp, &input.modulus);
+                let fourth_root_y = x.pop().ok_or_else(|| {
+                    error!("Expected to get a fourth root, but did not.");
+                    InternalError::InternalInvariantFailed
+                });
+                Ok(PiModProofElements {
+                    fourth_root: fourth_root_y?,
+                    sign_exponent: a,
+                    jacobi_exponent: b,
+                    challenge_secret_link: z,
+                    challenge: y,
+                })
+            })
+            .collect::<Result<Vec<PiModProofElements>>>();
+
+        let elements = elements?;
```

Comment:
The point of the previous comment was to pull up the `?` into the same line, instead of having it on a separate line. You also don't have to explicitly specify the type `PiModProofElements`, since the compiler can infer it.

Reply:
done.

Reply:
done.